src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionTextCache.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class ExpressionTextCache
+    {
+        private readonly ConcurrentDictionary<LambdaExpression, string> _cacheEntries  = 
+                new ConcurrentDictionary<LambdaExpression, string>();
```

Comment:
I suspect this is fairly fragile and doesn't reuse the `string`s as much as it could. Performs reference equality which means it (a) won't work in any case hoisting in `ExpressionRewriter` doesn't happen and (b) won't detect that the expressions in `LabelFor(m => FirstName)` and `TextBoxFor(m => FirstName)` have the same name.

Add a test looking up two separate but identical lambda expressions. If I'm correct, will likely find multiple entries in the dictionary (and will need to implement a simple `IEqualityComparer<LambdaExpression>`). Otherwise I'm good :smiley_cat: